src/Npgsql/NpgsqlParameterCollection.cs

```diff
@@ -19,21 +20,48 @@ public sealed class NpgsqlParameterCollection : DbParameterCollection, IList<Npg
 
         // Dictionary lookups for GetValue to improve performance
         Dictionary<string, int>? _lookup;
-        Dictionary<string, int>? _lookupIgnoreCase;
 
         /// <summary>
         /// Initializes a new instance of the NpgsqlParameterCollection class.
         /// </summary>
-        internal NpgsqlParameterCollection() => InvalidateHashLookups();
+        internal NpgsqlParameterCollection() {}
 
         /// <summary>
-        /// Invalidate the hash lookup tables.  This should be done any time a change
-        /// may throw the lookups out of sync with the list.
+        /// Revise the lookup. This should be done any time a change
+        /// may throw the lookup out of sync with the list.
         /// </summary>
-        internal void InvalidateHashLookups()
+        internal void ReflectListMutationInLookup(NpgsqlParameter parameter, int? index = 0)
         {
-            _lookup = null;
-            _lookupIgnoreCase = null;
+            if (_lookup is null)
+                return;
+
+            // We can't go through IndexOf as we may query the same lookup that is now out of sync.
+            if (index == null)
+            {
+                index = -1;
+                for (var i = 0; i < _internalList.Count; i++)
+                    if (string.Equals(parameter.TrimmedName, _internalList[i].TrimmedName, StringComparison.OrdinalIgnoreCase))
+                        index = i;
+            }
+
+            if (index == -1)
```

Comment:
I think this code path only happens from Remove, so maybe just move this out to there?